File: leet_code_75/my_python_solutions/can_place_flowers.py
import math
def canPlaceFlowers(flowerbed, n):
    # get the flower bed length
    flowerbed_length = len(flowerbed)
    # determine the maximum number of flowers that can be planted in a flower bed
    max_flowers = math.ceil(flowerbed_length / 2)
    # get the number of flowers to plant
    flowers_to_plant = n
    # set a count for the number of flowers in the flower bed
    current_num_flowers = 0
    # get the count of flowers in the flower bed
    for i in range(flowerbed_length):
        if flowerbed[i] == 1:
            current_num_flowers += 1
    # get the total number of flowers expected in the flower bed
    total_flowers = current_num_flowers + flowers_to_plant
    # if the total number of flowers exceeds max flowers return false
    if total_flowers > max_flowers:
        return False
    # base case of n == 0
    elif n == 0:
        return True
    # base case of flower bed length == 1
    elif flowerbed_length == 1 and n <= max_flowers:
        return not flowerbed[0]
    # other cases for flower bed length >= 2
    else:
        # make a copy of the flowerbed
        current_flowerbed = flowerbed
        # set a counter for the current flower section
        i = 0
        # as long as we remain inside the flowerbed, enter
        while i < flowerbed_length:
            # if there are flowers to plant, and we are at the start, and both the current and next flower sections are empty, then enter
            if (flowers_to_plant != 0) and (i == 0) and (current_flowerbed[i] == current_flowerbed[i + 1]) and (current_flowerbed[i] == 0):
                # update the current flowerbed
                current_flowerbed[i] = 1
                # decrement flowers to plant
                flowers_to_plant -= 1
                # skip over 2 sections of the flower bed
                i += 2
            # else if there are flowers to plant, and we are at the end, and both the current and previous flower sections are empty, then enter
            elif (flowers_to_plant != 0) and (i == flowerbed_length - 1) and (current_flowerbed[i] == current_flowerbed[i - 1]) and (current_flowerbed[i] == 0):
                # update the current flowerbed
                current_flowerbed[i] = 1
                # decrement flowers to plant
                flowers_to_plant -= 1
                # skip over 2 sections of the flower bed
                i += 2
            # else if there are flowers to plant, we are anywhere else between the beginning and end of the flower bed, and the previous-current-next 
            # flower sections are empty, then enter
            elif (flowers_to_plant != 0) and (i > 0) and (i < flowerbed_length - 1) and (current_flowerbed[i] == current_flowerbed[i - 1]) and (current_flowerbed[i] == current_flowerbed[i + 1]) and (current_flowerbed[i] == 0):
                # update the current flowerbed
                current_flowerbed[i] = 1
                # decrement flowers to plant
                flowers_to_plant -= 1
                # skip over 2 sections of the flower bed
                i += 2
            # else it is the case that the current flower section has a flower planted already, and cannot be planted or 
            # there are flowers both in the previous and/or next flower sections and the current section cannot be planted
            else:
                # skip over 1 section of the flower bed
                i += 1
        #print("Updated Flowerbed: ", current_flowerbed)
        #print("Flowers left to plant: ", flowers_to_plant)
        # return true if there are no more flowers to plant, else returns false
        return not flowers_to_plant

flowerbed = [1,0,0,0,1]
n = 1
print("Flowerbed:         ", flowerbed, " Flowers to plant = ", n)
print("Can be planted: ", canPlaceFlowers(flowerbed, n), " Expected: True\n")

flowerbed = [1,0,0,0,1]
n = 2
print("Flowerbed:         ", flowerbed, " Flowers to plant = ", n)
print("Can be planted: ", canPlaceFlowers(flowerbed, n), " Expected: False\n")

flowerbed = [0, 1, 0]
n = 1
print("Flowerbed:         ", flowerbed, " Flowers to plant = ", n)
print("Can be planted: ", canPlaceFlowers(flowerbed, n), " Expected: False\n")

flowerbed = [0,1,0,1,0,1,0,0]
n = 1
print("Flowerbed:         ", flowerbed, " Flowers to plant = ", n)
print("Can be planted: ", canPlaceFlowers(flowerbed, n), " Expected: True\n")

flowerbed = [0,0,1,0,0,0,0,1,0,1,0,0,0,1,0,0,1,0,1,0,1,0,0,0,1,0,1,0,1,0,0,1,0,0,0,0,0,1,0,1,0,0,0,1,0,0,1,0,0,0,1,0,0,1,0,0,1,0,0,0,1,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,1,0,1,0,0,0,0,0,0]
n = 17
print("Flowerbed:         ", flowerbed, " Flowers to plant = ", n)
print("Can be planted: ", canPlaceFlowers(flowerbed, n), " Expected False")

File: leet_code_75/my_java_solutions/CanPlaceFlowers.java
/**
    You have a long flowerbed in which some of the plots are planted, and some are not. However, flowers cannot be planted in adjacent plots.

    Given an integer array flowerbed containing 0's and 1's, where 0 means empty and 1 means not empty, and an integer n, return true if n new flowers can be planted in the flowerbed without violating the no-adjacent-flowers rule and false otherwise.

    Example 1:
    Input: flowerbed = [1,0,0,0,1], n = 1
    Output: true
    
    Example 2:
    Input: flowerbed = [1,0,0,0,1], n = 2
    Output: false
    
    Constraints:
    1 <= flowerbed.length <= 2 x 10^4
    flowerbed[i] is 0 or 1.
    There are no two adjacent flowers in flowerbed.
    0 <= n <= flowerbed.length
 */

import java.lang.Math;
class CanPlaceFlowers {
    /**
        This function determines whether flowers can be planted, without violating the contraints set by the problem
        @param flowerbed an array representing the current flowerbed consisting of 0's and 1's. (0 means empty) and (1 means flowered)
        @return true or false
    */
    public boolean canPlaceFlowers(int[] flowerbed, int n) {
        // get the flower bed length
        int flowerBedLength = flowerbed.length;
        // determine the maximum number of flowers that can be planted in a flower bed
        int maxFlowers = (int)(Math.ceil((double) flowerBedLength / 2));
        // get the number of flowers to plant
        int flowersToPlant = n;
        // set a count for the total number of flowers
        int currentNumFlowers = 0;
        // set a count for the total number of flowers currently in the flower bed
        for (int i = 0; i < flowerBedLength; i++) {
            if (flowerbed[i] == 1) {
                currentNumFlowers++;
            }
        }
        // get the total number of flowers expected in the flower bed
        int totalFlowers = currentNumFlowers + flowersToPlant;
        // base case, if too many flowers will be planted, return false
        if (totalFlowers > maxFlowers) {
            return false;
            // else if the flowers to plant is 0, you can always plant 0 flowers
        } else if (flowersToPlant == 0) {
            return true;
            // else if there is exactly one plot...
        } else if ((flowerBedLength == 1) && (flowersToPlant <= maxFlowers)) {
            if (flowerbed[0] == 0) {
                return true;
            } else {
                return false;
            }
            // else there are more than one flower plot
        } else {
            // create a shallow copy of the flowerbed
            int currentFlowerBed[] = flowerbed.clone();
            // set a counter for the current flower plot
            int i = 0;
            // as long as we remain in the flower bed, enter
            while (i < flowerBedLength) {
                // if there are no more flowers to plant, break out the loop
                if (flowersToPlant == 0) {
                    break;
                }
                // Case 1: we are at the start, and both the current and next flower plots are empty
                boolean firstPlotCase = (i == 0) && (currentFlowerBed[i] == currentFlowerBed[i + 1]) && (currentFlowerBed[i] == 0);
                // Case 2: we are anywhere else between the beginning and the end of the flower bed, and the 
                // previous-current-next flower plots are empty
                boolean middlePlotCase = (i > 0) && (i < (flowerBedLength - 1)) && (currentFlowerBed[i] == currentFlowerBed[i - 1]) && (currentFlowerBed[i] == currentFlowerBed[i + 1]) && (currentFlowerBed[i] == 0);
                // Case 3: we are at the end, and both the current and previous flower plots are empty
                boolean lastPlotCase = (i == flowerBedLength - 1) && (currentFlowerBed[i] == currentFlowerBed[i - 1]) && (currentFlowerBed[i] == 0);
                // set canPlantFlower to true if any of the cases above are true, else false
                boolean canPlantFlower = firstPlotCase || middlePlotCase || lastPlotCase;
                // if we can plant a flower, then plant it in the current flower plot
                if (canPlantFlower) {
                    currentFlowerBed[i] = 1;
                    flowersToPlant--;
                    i += 2;
                    // else skip over to the next flower plot
                } else {
                    i++;
                }
            }
        }
        // return true if there are no more flowers to plant, else false
        if (flowersToPlant == 0) {
            return true;
        } else {
            return false;
        }
    }

    /**
        This function prints the flowerBed, unmodified.
    */
    public void printFlowerBed(int[] flowerBed){
        System.out.print("[");
        int flowerBedLength = flowerBed.length;
        for(int i = 0; i < flowerBedLength; i++) {
            if(i == flowerBedLength - 1) {
                System.out.print(flowerBed[i]);
            } else {
                System.out.print(flowerBed[i] + ", ");
            }
       }
       System.out.println("]");
    }

    public static void main(String args[]){
        int flowerbed1[] = {1, 0, 0, 0, 1};
        int n = 1;
        CanPlaceFlowers canPlaceFlowers = new CanPlaceFlowers();
        canPlaceFlowers.printFlowerBed(flowerbed1);
        System.out.print("Can Place " + n + " Flowers? --> ");
        System.out.println(canPlaceFlowers.canPlaceFlowers(flowerbed1, n) + "\n");
        
        n = 2;
        canPlaceFlowers.printFlowerBed(flowerbed1);
        System.out.print("Can Place " + n + " Flowers? --> ");
        System.out.println(canPlaceFlowers.canPlaceFlowers(flowerbed1, n));

    }
}
